umbrella-feeds-reader/src/lib.rs

```diff
@@ -0,0 +1,266 @@
+    pub heartbeat: u64,
+    /// It is the time the validators run consensus to decide on the price data.
+    pub timestamp: Timestamp,
+    /// The price.
+    pub price: u128,
+}
+
+#[derive(Serial, Deserial, Debug, SchemaType)]
+struct State {
+    /// Registry contract where the list of all addresses of this protocol is stored.
+    registry: ContractAddress,
+    /// Umbrella_feeds contract where price data is stored.
+    umbrella_feeds: ContractAddress,
+    /// The key for the feed name represented by this contract. E.g. for the "ETH-USD" feed, the key will be the hash("ETH-USD").
+    key: HashSha2256,
+    /// Description of this feed name.
+    description: String,
+    /// Decimals for prices stored in the umbrella_feeds contract.
+    decimals: u8,
+}
+
+/// All smart contract errors.
+#[derive(Debug, PartialEq, Eq, Reject, Serial, SchemaType)]
+enum CustomContractError {
+    /// Failed to parse the parameter.
+    #[from(ParseError)]
+    ParseParams, // -1
+    /// Failed to invoke a contract.
+    InvokeContractError, // -2
+    /// Failed because the address(0x0) is not valid.
+    EmptyAddress, // -3
+    /// Failed because the decimal value in this contract and the decimal value in the umbrella_feeds contract do not match.
+    DecimalsDoesNotMatch, // -4
+}
+
+/// Mapping errors related to contract invocations to CustomContractError.
+impl<T> From<CallContractError<T>> for CustomContractError {
+    fn from(_cce: CallContractError<T>) -> Self {
+        Self::InvokeContractError
+    }
+}
+
+/// The parameter type for the contract init function.
+#[derive(Debug, Serialize, SchemaType)]
+pub struct InitParamsUmbrellaFeedsReader {
+    pub registry: ContractAddress,
+    pub umbrella_feeds: ContractAddress,
+    pub decimals: u8,
+    pub description: String,
+}
+
+/// Init function that creates a new smart contract. The `checkSetUp` entry point should be called after creating a new smart contract instance for a sanity check.
+#[init(
+    contract = "umbrella_feeds_reader",
+    parameter = "InitParamsUmbrellaFeedsReader",
+    crypto_primitives
+)]
+fn init<S: HasStateApi>(
+    ctx: &impl HasInitContext,
+    _state_builder: &mut StateBuilder<S>,
+    crypto_primitives: &impl HasCryptoPrimitives,
+) -> InitResult<State> {
+    let param: InitParamsUmbrellaFeedsReader = ctx.parameter_cursor().get()?;
+
+    ensure!(
+        param.registry
+            != ContractAddress {
+                index: 0,
+                subindex: 0
+            },
+        CustomContractError::EmptyAddress.into()
+    );
+
+    let key_hash = crypto_primitives
+        .hash_sha2_256(param.description.as_bytes())
+        .0;
+
+    Ok(State {
+        registry: param.registry,
+        decimals: param.decimals,
+        umbrella_feeds: param.umbrella_feeds,
+        key: HashSha2256(key_hash),
+        description: param.description,
+    })
+}
+
+/// View function to do a sanity check.
```

Comment:
This contract is provided to third-party dApp developers of umbrella. It can be deployed by them if they want to use it for easier integration into the umbrella ecosystem. Since third parties will deploy this contract, there were extra sanity checks in the constructor to make sure the input parameters/setup of the contract made sense. Since Concordium has no `host` function in the constructor and hence it is impossible to invoke anther contract form within the constructor, the only similar feature I was able to provide when translation the contracts was an extra entry point that can be called for sanity check.

@DZariusz